From 95e3a8d7db7b565f8fa782f73ab10659822d1c77 Mon Sep 17 00:00:00 2001
From: Maryam Tahhan <mtahhan@redhat.com>
Date: Wed, 18 May 2022 09:04:49 -0400
Subject: [PATCH] load custom bpf program -- shouldn't do on a per port basis
 -- need to just get the xskmap fd

Hacky implementation for POC... Should get back BPF Tool...

Signed-off-by: Maryam Tahhan <mtahhan@redhat.com>
---
 examples/cndpfwd/main.h                 |  16 ++--
 examples/cndpfwd/parse-args.c           |   1 +
 examples/cnet-graph/cnetfwd-graph.jsonc |  20 +++--
 examples/cnet-graph/parse-args.c        |   1 +
 examples/l3fwd-graph/parse-args.c       |   1 +
 lib/core/xskdev/xskdev.c                | 110 ++++++++++++++++++++++--
 lib/core/xskdev/xskdev.h                |  10 ++-
 lib/include/cne_lport.h                 |  33 +++----
 lib/usr/app/jcfg/jcfg.h                 |  22 ++---
 lib/usr/app/jcfg/jcfg_lport.c           |   5 +-
 tools/cndp.schema                       |  10 ++-
 11 files changed, 175 insertions(+), 54 deletions(-)

diff --git a/examples/cndpfwd/main.h b/examples/cndpfwd/main.h
index 1a5a8fa..844d671 100644
--- a/examples/cndpfwd/main.h
+++ b/examples/cndpfwd/main.h
@@ -49,12 +49,13 @@ enum {
     FWD_ACL_STATS   = (1 << 4), /**< Enable printing ACL stats */
 };
 
-#define PKT_API_TAG    "pkt_api"    /**< Packet API json tag */
-#define NO_METRICS_TAG "no-metrics" /**< json tag for no-metrics */
-#define NO_RESTAPI_TAG "no-restapi" /**< json tag for no-restapi */
-#define ENABLE_CLI_TAG "cli"        /**< json tag to enable/disable CLI */
-#define MODE_TAG       "mode"       /**< json tag to set the mode flag */
-#define UDS_PATH_TAG   "uds_path"   /**< json tag for UDS to get xsk map fd */
+#define PKT_API_TAG         "pkt_api"    /**< Packet API json tag */
+#define NO_METRICS_TAG      "no-metrics" /**< json tag for no-metrics */
+#define NO_RESTAPI_TAG      "no-restapi" /**< json tag for no-restapi */
+#define ENABLE_CLI_TAG      "cli"        /**< json tag to enable/disable CLI */
+#define MODE_TAG            "mode"       /**< json tag to set the mode flag */
+#define UDS_PATH_TAG        "uds_path"   /**< json tag for UDS to get xsk map fd */
+#define CUSTOM_BPF_PROG_TAG "custom-bpf-prog"
 
 #define MODE_DROP           "drop"           /**< Drop the received packets */
 #define MODE_RX_ONLY        "rx-only"        /**< Alias for MODE_DROP */
@@ -127,7 +128,8 @@ struct fwd_info {
     struct app_options opts;   /**< Application options*/
     pkt_api_t pkt_api;         /**< The packet API mode */
     uds_info_t *xdp_uds;       /**< UDS to get xsk map fd from */
-    int burst;                 /**< Burst Size */
+    char bpf_prog_path[64];
+    int burst; /**< Burst Size */
 };
 
 struct thread_func_arg_t {
diff --git a/examples/cndpfwd/parse-args.c b/examples/cndpfwd/parse-args.c
index ec14932..300d51b 100644
--- a/examples/cndpfwd/parse-args.c
+++ b/examples/cndpfwd/parse-args.c
@@ -183,6 +183,7 @@ process_callback(jcfg_info_t *j __cne_unused, void *_obj, void *arg, int idx)
             strlcpy(pcfg.pmd_name, lport->pmd_name, sizeof(pcfg.pmd_name));
             strlcpy(pcfg.ifname, lport->netdev, sizeof(pcfg.ifname));
             strlcpy(pcfg.name, lport->name, sizeof(pcfg.name));
+            strlcpy(pcfg.bpf_prog_path, lport->bpf_prog_path, sizeof(pcfg.bpf_prog_path));
 
             switch (f->pkt_api) {
             case XSKDEV_PKT_API:
diff --git a/examples/cnet-graph/cnetfwd-graph.jsonc b/examples/cnet-graph/cnetfwd-graph.jsonc
index 6f3552f..54a7dff 100644
--- a/examples/cnet-graph/cnetfwd-graph.jsonc
+++ b/examples/cnet-graph/cnetfwd-graph.jsonc
@@ -84,17 +84,21 @@
     "lports": {
         "eth0:0": {
             "pmd": "net_af_xdp",
-            "qid": 36,
+            "qid": 0,
             "umem": "umem0",
             "region": 0,
-            "description": "LAN 0 port"
+            "description": "LAN 0 port",
+            "skb_mode": true,
+            "custom_bpf_prog": "my_xdp_prog_kern.o"
         },
         "eth1:0": {
             "pmd": "net_af_xdp",
-            "qid": 37,
+            "qid": 0,
             "umem": "umem0",
             "region": 1,
-            "description": "LAN 1 port"
+            "description": "LAN 1 port",
+            "skb_mode": true,
+            "custom_bpf_prog": "my_xdp_prog_kern.o"
         }
     },
 
@@ -105,10 +109,10 @@
     // The initial group is for the main thread of the application.
     // The default group is special and is used if a thread if not assigned to a group.
     "lcore-groups": {
-        "initial": [35],
-        "group0": [38],
-        "group1": [39],
-        "default": ["38-39"]
+        "initial": [5],
+        "group0": [8],
+        "group1": [9],
+        "default": ["8-9"]
     },
 
     // (O) Set of common options application defined.
diff --git a/examples/cnet-graph/parse-args.c b/examples/cnet-graph/parse-args.c
index 2a14f40..b95ecf0 100644
--- a/examples/cnet-graph/parse-args.c
+++ b/examples/cnet-graph/parse-args.c
@@ -141,6 +141,7 @@ process_callback(jcfg_info_t *j, void *_obj, void *arg, int idx)
             strlcpy(pcfg.pmd_name, lport->pmd_name, sizeof(pcfg.pmd_name));
             strlcpy(pcfg.ifname, lport->netdev, sizeof(pcfg.ifname));
             strlcpy(pcfg.name, lport->name, sizeof(pcfg.name));
+            strlcpy(pcfg.bpf_prog_path, lport->bpf_prog_path, sizeof(pcfg.bpf_prog_path));
 
             pd->lport = pktdev_port_setup(&pcfg);
             if (pd->lport < 0) {
diff --git a/examples/l3fwd-graph/parse-args.c b/examples/l3fwd-graph/parse-args.c
index a777af8..beb5a6c 100644
--- a/examples/l3fwd-graph/parse-args.c
+++ b/examples/l3fwd-graph/parse-args.c
@@ -133,6 +133,7 @@ process_callback(jcfg_info_t *j __cne_unused, void *_obj, void *arg, int idx)
             strlcpy(pcfg.pmd_name, lport->pmd_name, sizeof(pcfg.pmd_name));
             strlcpy(pcfg.ifname, lport->netdev, sizeof(pcfg.ifname));
             strlcpy(pcfg.name, lport->name, sizeof(pcfg.name));
+            strlcpy(pcfg.bpf_prog_path, lport->bpf_prog_path, sizeof(pcfg.bpf_prog_path));
 
             pd->lport = pktdev_port_setup(&pcfg);
             if (pd->lport < 0) {
diff --git a/lib/core/xskdev/xskdev.c b/lib/core/xskdev/xskdev.c
index 9c934b0..527cabe 100644
--- a/lib/core/xskdev/xskdev.c
+++ b/lib/core/xskdev/xskdev.c
@@ -789,6 +789,80 @@ xskdev_buf_set_buf_mgmt_ops(lport_buf_mgmt_t *dst, lport_buf_mgmt_t *src)
         memcpy(dst, src, sizeof(lport_buf_mgmt_t));
 }
 
+static int
+load_program(const char *prog_path, struct bpf_object **obj)
+{
+    int ret, prog_fd;
+
+    ret = bpf_prog_load(prog_path, BPF_PROG_TYPE_XDP, obj, &prog_fd);
+    if (ret)
+        return -1;
+
+    return prog_fd;
+}
+
+static int
+xskdev_load_custom_xdp_prog(const char *prog_path, int if_index, struct bpf_map **map, __u32 flags)
+{
+    int ret, prog_fd;
+    struct bpf_object *obj;
+
+    prog_fd = load_program(prog_path, &obj);
+    if (prog_fd < 0) {
+        CNE_ERR("Failed to load program %s\n", prog_path);
+        return -1;
+    }
+
+    /*
+     * The loaded program must provision for a map of xsks, such that some
+     * traffic can be redirected to userspace.
+     */
+    *map = bpf_object__find_map_by_name(obj, "xsks_map");
+    if (!*map) {
+        CNE_ERR("Failed to find xsks_map in %s\n", prog_path);
+        return -1;
+    }
+
+    /* Link the program with the given network device */
+    ret = bpf_set_link_xdp_fd(if_index, prog_fd, flags);
+    if (ret == -EEXIST && !(flags & XDP_FLAGS_UPDATE_IF_NOEXIST)) {
+        /* Force mode didn't work, probably because a program of the
+         * opposite type is loaded. Let's unload that and try loading
+         * again.
+         */
+
+        __u32 old_flags = flags;
+
+        flags &= ~XDP_FLAGS_MODES;
+        flags |= (old_flags & XDP_FLAGS_SKB_MODE) ? XDP_FLAGS_DRV_MODE : XDP_FLAGS_SKB_MODE;
+        ret = bpf_set_link_xdp_fd(if_index, -1, flags);
+        if (!ret)
+            ret = bpf_set_link_xdp_fd(if_index, prog_fd, old_flags);
+    }
+    if (ret < 0) {
+        CNE_ERR("Failed to set prog fd %d on interface\n", prog_fd);
+        switch (-ret) {
+        case EBUSY:
+        case EEXIST:
+            CNE_ERR("Hint: XDP already loaded on device"
+                    " use --force to swap/replace\n");
+            break;
+        case EOPNOTSUPP:
+            CNE_ERR("Hint: Native-XDP not supported"
+                    " use --skb-mode or --auto-mode\n");
+            break;
+        default:
+            CNE_ERR("UNKOWN ERROR ret = %d\n", ret);
+            break;
+        }
+        return -1;
+    }
+
+    CNE_INFO("Successfully loaded XDP program %s with fd %d\n", prog_path, prog_fd);
+
+    return 0;
+}
+
 xskdev_info_t *
 xskdev_socket_create(struct lport_cfg *c)
 {
@@ -821,6 +895,26 @@ xskdev_socket_create(struct lport_cfg *c)
             CNE_ERR_GOTO(err, "Failed to receive xsk map fd\n");
     }
 
+    if (c->flags & LPORT_CUST_BPF_PROG) {
+        __u32 xdp_flags = 0;
+
+        if (!c->bpf_prog_path)
+            CNE_ERR_GOTO(err, "bpf_prog_path EMPTY");
+
+        xi->load_custom_bpf_prog = true;
+
+        xdp_flags &= ~XDP_FLAGS_MODES; /* Clear flags */
+        xdp_flags |= XDP_FLAGS_SKB_MODE;
+
+        // LOAD PROGRAM, get xi->xsk_map_fd
+        ret = xskdev_load_custom_xdp_prog(c->bpf_prog_path, if_index, &xi->xsk_map, xdp_flags);
+        if (ret) {
+            CNE_ERR_GOTO(err, "Failed to load_custom_xdp_prog\n");
+        }
+        cfg.libbpf_flags |= XSK_LIBBPF_FLAGS__INHIBIT_PROG_LOAD;
+        xi->xsk_map_fd = bpf_map__fd(xi->xsk_map);
+    }
+
     if (xskdev_use_tx_lock) {
         ret = pthread_mutex_init(&xi->tx_lock, NULL);
         if (ret)
@@ -910,13 +1004,13 @@ xskdev_socket_create(struct lport_cfg *c)
     xi->skb_mode     = (c->flags & LPORT_SKB_MODE) ? true : false;
     xi->shared_umem  = (c->flags & LPORT_SHARED_UMEM) ? true : false;
 
-    xi->xdp_flags    = XDP_FLAGS_UPDATE_IF_NOEXIST;
-    xi->xdp_flags    = ((xi->skb_mode) ? XDP_FLAGS_SKB_MODE : XDP_FLAGS_DRV_MODE) | xi->xdp_flags;
-    cfg.xdp_flags    = xi->xdp_flags;
-    cfg.bind_flags   = (xi->skb_mode) ? XDP_COPY | XDP_USE_NEED_WAKEUP : XDP_USE_NEED_WAKEUP;
-    cfg.rx_size      = c->rx_nb_desc;
-    cfg.tx_size      = c->tx_nb_desc;
-    cfg.libbpf_flags = xi->unprivileged ? XSK_LIBBPF_FLAGS__INHIBIT_PROG_LOAD : 0;
+    xi->xdp_flags  = XDP_FLAGS_UPDATE_IF_NOEXIST;
+    xi->xdp_flags  = ((xi->skb_mode) ? XDP_FLAGS_SKB_MODE : XDP_FLAGS_DRV_MODE) | xi->xdp_flags;
+    cfg.xdp_flags  = xi->xdp_flags;
+    cfg.bind_flags = (xi->skb_mode) ? XDP_COPY | XDP_USE_NEED_WAKEUP : XDP_USE_NEED_WAKEUP;
+    cfg.rx_size    = c->rx_nb_desc;
+    cfg.tx_size    = c->tx_nb_desc;
+    cfg.libbpf_flags |= xi->unprivileged ? XSK_LIBBPF_FLAGS__INHIBIT_PROG_LOAD : 0;
 
     if (xi->busy_polling) {
         xi->busy_budget  = (c->busy_budget) ? c->busy_budget : AF_XDP_DFLT_BUSY_BUDGET;
@@ -949,7 +1043,7 @@ xskdev_socket_create(struct lport_cfg *c)
     xi->txq.fds.fd     = xsk_socket__fd(xi->rxq.xsk);
     xi->txq.fds.events = POLLIN;
 
-    if (xi->unprivileged) {
+    if (xi->unprivileged || xi->load_custom_bpf_prog) {
         ret = xsk_socket__update_xskmap(xi->rxq.xsk, xi->xsk_map_fd);
         if (ret)
             CNE_ERR_GOTO(err, "Update of BPF map failed. %s\n", strerror(errno));
diff --git a/lib/core/xskdev/xskdev.h b/lib/core/xskdev/xskdev.h
index 29b44e4..9b6d84f 100644
--- a/lib/core/xskdev/xskdev.h
+++ b/lib/core/xskdev/xskdev.h
@@ -115,13 +115,15 @@ typedef struct xskdev_info {
     uint32_t busy_budget;          /**< Busy polling budget value */
     uds_info_t *uds_info;          /**< UDS info struct */
     int xsk_map_fd;                /**< xsk map file descriptor from UDS */
+    struct bpf_map *xsk_map;
 
     /* byte flags to mirror the lport_cfg_t.flags bits */
     bool unprivileged; /**< Inhibit privileged ops (BPF program load & config of busy poll) */
-    bool needs_wakeup; /**< Force the lport to use wakeup calls */
-    bool skb_mode;     /**< Force lport to use SKB Copy mode */
-    bool busy_polling; /**< Enable the lport to use busy polling if available */
-    bool shared_umem;  /**< Enable Shared UMEM support */
+    bool load_custom_bpf_prog; /**< load a custom BPF program) */
+    bool needs_wakeup;         /**< Force the lport to use wakeup calls */
+    bool skb_mode;             /**< Force lport to use SKB Copy mode */
+    bool busy_polling;         /**< Enable the lport to use busy polling if available */
+    bool shared_umem;          /**< Enable Shared UMEM support */
 
     lport_buf_mgmt_t buf_mgmt; /**< Buffer management routines structure */
     xskdev_get_mbuf_addr_tx_t
diff --git a/lib/include/cne_lport.h b/lib/include/cne_lport.h
index e0d2121..05eab88 100644
--- a/lib/include/cne_lport.h
+++ b/lib/include/cne_lport.h
@@ -71,21 +71,23 @@ typedef struct lport_cfg {
     char name[LPORT_NAME_LEN];     /**< logical port name */
     char ifname[LPORT_NAME_LEN];   /**< Interface name or netdev name */
     char pmd_name[LPORT_NAME_LEN]; /**< Name of the PMD i.e. net_af_xdp, net_ring */
-    uint16_t flags;                /**< Flags to configure the AF_XDP interface */
-    uint16_t qid;                  /**< Queue ID */
-    uint32_t bufcnt;               /**< Number of buffers in the pool */
-    uint32_t bufsz;                /**< Size of the buffers in the UMEM space */
-    uint32_t rx_nb_desc;           /**< Number of RX descriptor entries */
-    uint32_t tx_nb_desc;           /**< Number of TX descriptor entries */
-    uint16_t busy_timeout;         /**< 1-65535 or 0 - use default value, value in milliseconds */
-    uint16_t busy_budget;          /**< -1 disabled, 0 use default, >0 budget value */
-    void *addr;                    /**< Start address of the buffers */
-    char *umem_addr;               /**< Address of the allocated UMEM area */
-    char *pmd_opts;                /**< options string from jasonc file */
-    size_t umem_size;              /**< Size of the umem region */
-    pktmbuf_info_t *pi;            /**< pktmbuf_info_t structure pointer */
-    void *xsk_uds;                 /**< The UDS to connect to get xsk FDs */
-    lport_buf_mgmt_t buf_mgmt;     /**< Buffer management functions */
+    char bpf_prog_path[64];
+    uint16_t flags;            /**< Flags to configure the AF_XDP interface */
+    uint16_t qid;              /**< Queue ID */
+    uint32_t bufcnt;           /**< Number of buffers in the pool */
+    uint32_t bufsz;            /**< Size of the buffers in the UMEM space */
+    uint32_t rx_nb_desc;       /**< Number of RX descriptor entries */
+    uint32_t tx_nb_desc;       /**< Number of TX descriptor entries */
+    uint16_t busy_timeout;     /**< 1-65535 or 0 - use default value, value in milliseconds */
+    uint16_t busy_budget;      /**< -1 disabled, 0 use default, >0 budget value */
+    void *addr;                /**< Start address of the buffers */
+    char *umem_addr;           /**< Address of the allocated UMEM area */
+    char *pmd_opts;            /**< options string from jasonc file */
+    size_t umem_size;          /**< Size of the umem region */
+    pktmbuf_info_t *pi;        /**< pktmbuf_info_t structure pointer */
+    void *xsk_uds;             /**< The UDS to connect to get xsk FDs */
+    lport_buf_mgmt_t buf_mgmt; /**< Buffer management functions */
+
 } lport_cfg_t;
 
 /**< lport_cfg.flags configuration bits */
@@ -96,6 +98,7 @@ typedef struct lport_cfg {
 #define LPORT_SHARED_UMEM            (1 << 4) /**< Enable UMEM Shared mode if available */
 #define LPORT_USER_MANAGED_BUFFERS   (1 << 5) /**< Enable Buffer Manager outside of CNDP */
 #define LPORT_UMEM_UNALIGNED_BUFFERS (1 << 6) /**< Enable unaligned frame UMEM support */
+#define LPORT_CUST_BPF_PROG          (1 << 7)
 
 typedef struct lport_stats {
     uint64_t ipackets;           /**< Total number of successfully received packets. */
diff --git a/lib/usr/app/jcfg/jcfg.h b/lib/usr/app/jcfg/jcfg.h
index bab38df..6615973 100644
--- a/lib/usr/app/jcfg/jcfg.h
+++ b/lib/usr/app/jcfg/jcfg.h
@@ -43,15 +43,16 @@ struct json_object;
 /**
  * The standard list of section names used in  JCFG json-c files.
  */
-#define APP_TAG         "application"
-#define DEFAULT_TAG     "defaults"
-#define OPTION_TAG      "options"
-#define UMEM_TAG        "umems"
-#define LPORT_TAG       "lports"
-#define LGROUP_TAG      "lcore-groups"
-#define THREAD_TAG      "threads"
-#define LPORT_GROUP_TAG "lport-groups"
-#define USER_TAG        "users"
+#define APP_TAG             "application"
+#define DEFAULT_TAG         "defaults"
+#define OPTION_TAG          "options"
+#define UMEM_TAG            "umems"
+#define LPORT_TAG           "lports"
+#define LGROUP_TAG          "lcore-groups"
+#define THREAD_TAG          "threads"
+#define LPORT_GROUP_TAG     "lport-groups"
+#define USER_TAG            "users"
+#define CUSTOM_BPF_PROG_TAG "custom-bpf-prog"
 
 /**
  * Macro to initialize a const char *tags[] type array for indexing with jcfg_tag_t.
@@ -174,6 +175,7 @@ typedef struct jcfg_lport {
     uint16_t busy_timeout;       /**< busy timeout value in milliseconds */
     uint16_t busy_budget;        /**< busy budget 0xFFFF disabled, 0 use default, >0 budget */
     uint16_t flags; /**< Flags to configure lport in lport_cfg_t.flags in cne_lport.h */
+    char *bpf_prog_path;
 } jcfg_lport_t;
 
 /** JCFG lport configuration names */
@@ -190,7 +192,7 @@ typedef struct jcfg_lport {
 #define JCFG_LPORT_UNPRIVILEGED_NAME "unprivileged"
 #define JCFG_LPORT_FORCE_WAKEUP_NAME "force_wakeup"
 #define JCFG_LPORT_SKB_MODE_NAME     "skb_mode"
-
+#define JCFG_CUST_BPF_PROG_NAME      "custom_bpf_prog"
 /**
  * JCFG  lgroup for lcore allocations
  */
diff --git a/lib/usr/app/jcfg/jcfg_lport.c b/lib/usr/app/jcfg/jcfg_lport.c
index e60c751..f69b14d 100644
--- a/lib/usr/app/jcfg/jcfg_lport.c
+++ b/lib/usr/app/jcfg/jcfg_lport.c
@@ -94,7 +94,10 @@ _lport(struct json_object *obj, int flags, struct json_object *parent __cne_unus
         else if (!strncmp(key, JCFG_LPORT_DESC_NAME, keylen) ||
                  !strncmp(key, JCFG_LPORT_DESCRIPTION_NAME, keylen))
             lport->desc = strdup(json_object_get_string(obj));
-        else if (!strncmp(key, JCFG_LPORT_BUSY_TIMEOUT_NAME, keylen))
+        else if (!strncmp(key, JCFG_CUST_BPF_PROG_NAME, keylen)) {
+            lport->bpf_prog_path = strndup(json_object_get_string(obj), JCFG_MAX_STRING_SIZE);
+            lport->flags |= LPORT_CUST_BPF_PROG;
+        } else if (!strncmp(key, JCFG_LPORT_BUSY_TIMEOUT_NAME, keylen))
             lport->busy_timeout = (uint16_t)json_object_get_int(obj);
         else if (!strncmp(key, JCFG_LPORT_BUSY_BUDGET_NAME, keylen))
             lport->busy_budget = (uint16_t)json_object_get_int(obj);
diff --git a/tools/cndp.schema b/tools/cndp.schema
index ec4f00b..f7573c4 100644
--- a/tools/cndp.schema
+++ b/tools/cndp.schema
@@ -166,7 +166,11 @@
                             "description": "Enable XDP_FLAGS_SKB_MODE to force copy mode",
                             "type": "boolean",
                             "default": false
-                        }
+                        },
+                    "custom_bpf_prog": {
+                        "description": "The custom bpf program to load",
+                        "type": "string"
+                    }
                     },
                     "patternProperties": {
                         "^desc(ription){0,1}$": {
@@ -344,6 +348,10 @@
                         "description": "Enable XDP_FLAGS_SKB_MODE to force copy mode",
                         "type": "boolean",
                         "default": false
+                    },
+                    "custom_bpf_prog": {
+                        "description": "The custom bpf program to load",
+                        "type": "string"
                     }
                 },
                 "patternProperties": {
-- 
2.35.1

